File: typescript/nodes/startNode.test.ts
import { assertEquals } from "@std/assert";
import * as StartNode from "./startNode.ts";

Deno.test("startNode should transition to onSuccess with default count when input is empty", () => {
  const state = { count: 0 };
  const utils: StartNode.Utils = {
    prompt: () => "",
    error: () => {},
  };
  const node = StartNode.create({ onSuccess: "count", onError: "start" }, utils);
  const [next, nextState] = node(state, "start") as [string, { count: number }];
  assertEquals(next, "count");
  assertEquals(nextState.count, 0);
});

Deno.test("startNode should transition to onSuccess with parsed count when input is valid", () => {
  const state = { count: 0 };
  const utils: StartNode.Utils = {
    prompt: () => "42",
    error: () => {},
  };
  const node = StartNode.create({ onSuccess: "count", onError: "start" }, utils);
  const [next, nextState] = node(state, "start") as [string, { count: number }];
  assertEquals(next, "count");
  assertEquals(nextState.count, 42);
});

Deno.test("startNode should transition to onError when input is invalid", () => {
  const state = { count: 10 };
  const utils: StartNode.Utils = {
    prompt: () => "invalid",
    error: () => {},
  };
  const node = StartNode.create({ onSuccess: "count", onError: "start" }, utils);
  const [next, nextState] = node(state, "start") as [string, { count: number }];
  assertEquals(next, "start");
  assertEquals(nextState.count, 10);
});

File: go/nodes/start_node_test.go
package nodes

import (
	"context"
	"testing"

	"github.com/argenkiwi/ambler/go"
)

type mockStartNodeUtils struct {
	input string
	err   error
}

func (m mockStartNodeUtils) Prompt(message string) (string, error) {
	return m.input, m.err
}

func (m mockStartNodeUtils) Error(message string) {}

func TestStartNode(t *testing.T) {
	ctx := context.Background()
	edges := StartNodeEdges[string]{
		OnSuccess: ambler.Ptr("count"),
		OnError:   ambler.Ptr("start"),
	}

	t.Run("empty input defaults to 0", func(t *testing.T) {
		utils := mockStartNodeUtils{input: ""}
		node := CreateStartNode[mockState, string](edges, utils)
		state := mockState{count: 10}

		next, err := node(ctx, state, "start")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}

		if next.NodeID == nil || *next.NodeID != "count" {
			t.Errorf("expected count, got %v", next.NodeID)
		}
		if next.State.GetCount() != 0 {
			t.Errorf("expected 0, got %d", next.State.GetCount())
		}
	})

	t.Run("valid input updates count", func(t *testing.T) {
		utils := mockStartNodeUtils{input: "42"}
		node := CreateStartNode[mockState, string](edges, utils)
		state := mockState{count: 10}

		next, err := node(ctx, state, "start")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}

		if next.NodeID == nil || *next.NodeID != "count" {
			t.Errorf("expected count, got %v", next.NodeID)
		}
		if next.State.GetCount() != 42 {
			t.Errorf("expected 42, got %d", next.State.GetCount())
		}
	})

	t.Run("invalid input transitions to error", func(t *testing.T) {
		utils := mockStartNodeUtils{input: "invalid"}
		node := CreateStartNode[mockState, string](edges, utils)
		state := mockState{count: 10}

		next, err := node(ctx, state, "start")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}

		if next.NodeID == nil || *next.NodeID != "start" {
			t.Errorf("expected start, got %v", next.NodeID)
		}
		if next.State.GetCount() != 10 {
			t.Errorf("expected 10, got %d", next.State.GetCount())
		}
	})
}
